### PR TITLE
SliverVisibility

### LINKED ISSUE #45182: Add SliverVisibility
A sliver compatible version of the Visibility widget would be nice to have.

This will require sliver compatible versions of 
- [x] Opacity #44289
- [x] IgnorePointer #45127
- [x] Offstage #45128

### PR DESCRIPTION
**Pull request by Piinks**
## Description

Adds the SliverVisibility widget!

## Related Issues

Fixes #45182

## Tests

SliverVisibility tests - oodles of them!

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by Piinks** [comment]
Blocked by #45580

**Comment by Piinks** [comment]
No longer blocked by SliverOffstage fix. :)

File: packages/flutter/lib/src/widgets/visibility.dart [hunk 1]
```diff
@@ -66,10 +68,22 @@ class Visibility extends StatelessWidget {
        assert(maintainState != null),
        assert(maintainAnimation != null),
        assert(maintainSize != null),
-       assert(maintainState == true || maintainAnimation == false, 'Cannot maintain animations if the state is not also maintained.'),
-       assert(maintainAnimation == true || maintainSize == false, 'Cannot maintain size if animations are not maintained.'),
-       assert(maintainSize == true || maintainSemantics == false, 'Cannot maintain semantics if size is not maintained.'),
-       assert(maintainSize == true || maintainInteractivity == false, 'Cannot maintain interactivity if size is not maintained.'),
+       assert(
+        maintainState == true || maintainAnimation == false,
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/visibility.dart]: nit: indentation with these asserts is incorrect.

File: packages/flutter/lib/src/widgets/visibility.dart [hunk 2]
```diff
@@ -243,3 +257,239 @@ class Visibility extends StatelessWidget {
...
+///  * [SliverOffstage], which can stop its sliver child from being laid out or
+///    painted.
+///  * [TickerMode], which can stop its child from being animated.
+///  * [ExcludeSemantics], which can hide the child from accessibility tools.
+///  * [SliverIgnorePointer], which can disable touch interactions with the
+///    sliver child.
+///
+/// Using this widget is not necessary to hide children. The simplest way to
+/// hide a child is just to not include it, or, if a child _must_ be given (e.g.
+/// because the parent is a [StatelessWidget]) then to use
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/visibility.dart]: This sentence is not complete.

File: packages/flutter/lib/src/widgets/visibility.dart [hunk 3]
```diff
@@ -243,3 +257,239 @@ class Visibility extends StatelessWidget {
...
+// TODO(Piinks): What is the default replacement here?
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/visibility.dart]: What do you mean by this TODO?

File: packages/flutter/lib/src/widgets/visibility.dart [hunk 4]
```diff
@@ -243,3 +257,239 @@ class Visibility extends StatelessWidget {
...
+  /// set if [maintainSize] is set.
+  ///
+  /// The [maintainSize] argument can only be set if [maintainAnimation] is set.
+  ///
+  /// The [maintainAnimation] argument can only be set if [maintainState] is
+  /// set.
+  const SliverVisibility({
+    Key key,
+    @required this.sliver,
+    this.replacementSliver = const SliverToBoxAdapter(child: SizedBox.shrink()),
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/visibility.dart]: Do we need a SizedSliver?

File: packages/flutter/lib/src/widgets/visibility.dart [hunk 5]
```diff
@@ -243,3 +257,239 @@ class Visibility extends StatelessWidget {
...
+    this.visible = true,
+    this.maintainState = false,
+    this.maintainAnimation = false,
+    this.maintainSize = false,
+    this.maintainSemantics = false,
+    this.maintainInteractivity = false,
+  }) : assert(sliver != null),
+      assert(visible != null),
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/visibility.dart]: nit: indentation.

File: packages/flutter/lib/src/widgets/visibility.dart [hunk 6]
```diff
@@ -243,3 +257,239 @@ class Visibility extends StatelessWidget {
...
+
+  /// The sliver to show or hide, as controlled by [visible].
+  ///
+  /// {@macro flutter.widgets.child}
+  final Widget sliver;
+
+  /// The widget to use when the sliver child is not [visible], assuming that
+  /// none of the `maintain` flags (in particular, [maintainState]) are set.
+  ///
+  /// The normal behavior is to replace the widget with a
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/visibility.dart]: Same here.

File: packages/flutter/test/widgets/visibility_test.dart [hunk 7]
```diff
@@ -310,4 +432,458 @@ void main() {
 
     semantics.dispose();
   }, skip: isBrowser);
+
+  testWidgets('SliverVisibility', (WidgetTester tester) async {
```
Comment (goderbauer) [inline packages/flutter/test/widgets/visibility_test.dart]: nit: I would probably put these in their own sliver_visiblity_test.dart file.

File: packages/flutter/lib/src/widgets/visibility.dart (hunk 3 above)
Comment (Piinks) [reply packages/flutter/lib/src/widgets/visibility.dart]: OH! I seem to have missed this in my clean up pass. Thanks!

File: packages/flutter/lib/src/widgets/visibility.dart (hunk 4 above)
Comment (Piinks) [reply packages/flutter/lib/src/widgets/visibility.dart]: I am curious if we do. If so, would it take boxes or slivers as children? What would be the general use case for such a sliver, beyond this one instance?

Comment (Piinks) [reply packages/flutter/lib/src/widgets/visibility.dart]: It turns out that SliverToBoxAdapter defaults to SliverGeometry.zero if given no child. I think that sounds like it makes sense here then without a child.

File: packages/flutter/test/widgets/visibility_test.dart (hunk 7 above)
Comment (Piinks) [reply packages/flutter/test/widgets/visibility_test.dart]: SGTM!

File: packages/flutter/lib/src/widgets/visibility.dart [hunk 8]
```diff
@@ -243,3 +257,238 @@ class Visibility extends StatelessWidget {
...
+    @required this.sliver,
+    this.replacementSliver = const SliverToBoxAdapter(),
+    this.visible = true,
+    this.maintainState = false,
+    this.maintainAnimation = false,
+    this.maintainSize = false,
+    this.maintainSemantics = false,
+    this.maintainInteractivity = false,
+  }) : assert(sliver != null),
+       assert(visible != null),
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/visibility.dart]: maintainSemantics, maintainInteractivity, and replacementSliver are missing an assert? (Or the doc claiming that they must not be null is incorrect).

File: packages/flutter/test/widgets/visibility_test.dart [hunk 9]
```diff
@@ -2,6 +2,8 @@
 // Use of this source code is governed by a BSD-style license that can be
 // found in the LICENSE file.
 
+import 'package:flutter/material.dart';
```
Comment (goderbauer) [inline packages/flutter/test/widgets/visibility_test.dart]: why these extra imports? As far as I can tell this file only has formatting changes?

File: packages/flutter/test/widgets/sliver_visibility_test.dart [hunk 10]
```diff
@@ -0,0 +1,490 @@
...
+    expect(renderSliver.constraints.crossAxisExtent, 800.0);
+    expect(semantics.nodesWith(label: 'a true'), hasLength(0));
+    expect(log, <String>['created new state']);
+    expect(find.byKey(anchor), findsNothing);
+    log.clear();
+
+    // Now we toggle the visibility off and on a few times to make sure that
+    // works.
+
+    // visible: true, maintain state/
```
Comment (goderbauer) [inline packages/flutter/test/widgets/sliver_visibility_test.dart]: nit: why the trailing /?

**Review by goderbauer — APPROVED** [review]
LGTM

**Pull request merged**